File: flutter_optical_transfer/lib/core/protocol.dart
import 'dart:typed_data';
import 'dart:convert';
import 'package:crypto/crypto.dart';
import 'package:archive/archive.dart';

const int headerLen = 20;
const int maxFileBytes = 1024 * 1024 * 1024; // 1 GB max (1024 MB)
const int fileHeaderLen = 49;
const int magic0 = 0xD1;
const int magic1 = 0x0C;
final Uint8List fileMagic = Uint8List.fromList([0x44, 0x43, 0x46, 0x32]); // "DCF2"

enum CompressionMode { none, gzip }

class PackedOpticalFile {
  final Uint8List container;
  final CompressionMode compression;
  final int originalSize;
  final int transmittedSize;

  PackedOpticalFile({
    required this.container,
    required this.compression,
    required this.originalSize,
    required this.transmittedSize,
  });
}

class OpticalFile {
  final String name;
  final String type;
  final Uint8List bytes;
  final Uint8List sha256;
  final CompressionMode compression;
  final int transmittedSize;

  OpticalFile({
    required this.name,
    required this.type,
    required this.bytes,
    required this.sha256,
    required this.compression,
    required this.transmittedSize,
  });
}

/// Calculate FNV-1a hash matching protocol.ts
int fnv1a(Uint8List bytes) {
  int hash = 0x811c9dc5;
  for (int i = 0; i < bytes.length; i++) {
    hash ^= bytes[i];
    // 32-bit integer multiplication 0x01000193 = 16777619
    hash = (hash * 16777619) & 0xFFFFFFFF;
  }
  return hash;
}

/// Calculate SHA-256 hash
Uint8List calculateSha256(Uint8List bytes) {
  return Uint8List.fromList(sha256.convert(bytes).bytes);
}

/// Sanitize file name for safe filesystem handling
String safeFileName(String name) {
  final base = name.split(RegExp(r'[/\\]')).last;
  final cleaned = base.replaceAll(RegExp(r'[\u0000-\u001f\u007f]'), '').trim();
  if (cleaned.isEmpty || cleaned == '.' || cleaned == '..') {
    return 'transfer.bin';
  }
  return cleaned;
}

/// Pack file into DCF2 container format
PackedOpticalFile packFile(String name, String type, Uint8List bytes) {
  if (bytes.isEmpty) throw Exception('Choose a non-empty file.');
  if (bytes.length > maxFileBytes) {
    throw Exception('Files are limited to 1 GB.');
  }

  final nameBytes = utf8.encode(safeFileName(name));
  final typeBytes = utf8.encode(type.isEmpty ? 'application/octet-stream' : type);
  
  if (nameBytes.length > 0xFFFF || typeBytes.length > 0xFFFF) {
    throw Exception('The file name or media type is too long.');
  }

  // Attempt GZIP compression if data is >= 768 bytes
  bool useGzip = false;
  Uint8List transmitted = bytes;
  CompressionMode mode = CompressionMode.none;

  if (bytes.length >= 768) {
    try {
      final compressed = Uint8List.fromList(GZipEncoder().encode(bytes)!);
      if (compressed.length + 64 < bytes.length) {
        transmitted = compressed;
        useGzip = true;
        mode = CompressionMode.gzip;
      }
    } catch (_) {
      transmitted = bytes;
    }
  }

  final sha = calculateSha256(bytes);
  final out = Uint8List(fileHeaderLen + nameBytes.length + typeBytes.length + transmitted.length);
  final bd = ByteData.sublistView(out);

  out.setRange(0, 4, fileMagic);
  bd.setUint8(4, useGzip ? 1 : 0);
  bd.setUint16(5, nameBytes.length, Endian.little);
  bd.setUint16(7, typeBytes.length, Endian.little);
  bd.setUint32(9, bytes.length, Endian.little);
  bd.setUint32(13, transmitted.length, Endian.little);
  out.setRange(17, 49, sha);

  int offset = fileHeaderLen;
  out.setRange(offset, offset + nameBytes.length, nameBytes);
  offset += nameBytes.length;

  out.setRange(offset, offset + typeBytes.length, typeBytes);
  offset += typeBytes.length;

  out.setRange(offset, offset + transmitted.length, transmitted);

  return PackedOpticalFile(
    container: out,
    compression: mode,
    originalSize: bytes.length,
    transmittedSize: transmitted.length,
  );
}

/// Unpack DCF2 container payload
OpticalFile unpackFile(Uint8List container) {
  if (container.length < fileHeaderLen) {
    throw Exception('Container payload is too small.');
  }

  for (int i = 0; i < 4; i++) {
    if (container[i] != fileMagic[i]) {
      throw Exception('Invalid file container header.');
    }
  }

  final bd = ByteData.sublistView(container);
  final useGzip = bd.getUint8(4) != 0;
  final nameLen = bd.getUint16(5, Endian.little);
  final typeLen = bd.getUint16(7, Endian.little);
  final declOriginalSize = bd.getUint32(9, Endian.little);
  final declTransmittedSize = bd.getUint32(13, Endian.little);

  final sha256Expected = container.sublist(17, 49);

  int offset = fileHeaderLen;
  if (offset + nameLen + typeLen + declTransmittedSize > container.length) {
    throw Exception('Container truncated or corrupted.');
  }

  final nameBytes = container.sublist(offset, offset + nameLen);
  offset += nameLen;

  final typeBytes = container.sublist(offset, offset + typeLen);
  offset += typeLen;

  final payloadBytes = container.sublist(offset, offset + declTransmittedSize);

  final name = safeFileName(utf8.decode(nameBytes, allowMalformed: true));
  final type = utf8.decode(typeBytes, allowMalformed: true);

  Uint8List decompressed;
  if (useGzip) {
    decompressed = Uint8List.fromList(GZipDecoder().decodeBytes(payloadBytes));
  } else {
    decompressed = payloadBytes;
  }

  if (decompressed.length != declOriginalSize) {
    throw Exception('Decompressed length mismatch.');
  }

  final actualSha = calculateSha256(decompressed);
  for (int i = 0; i < 32; i++) {
    if (actualSha[i] != sha256Expected[i]) {
      throw Exception('SHA-256 integrity check failed!');
    }
  }

  return OpticalFile(
    name: name,
    type: type,
    bytes: decompressed,
    sha256: actualSha,
    compression: useGzip ? CompressionMode.gzip : CompressionMode.none,
    transmittedSize: payloadBytes.length,
  );
}

File: android/app/src/main/java/com/example/opticaltransfer/core/protocol/Protocol.kt
package com.example.opticaltransfer.core.protocol

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

const val HEADER_LEN = 20
const val MAX_FILE_BYTES = 1024 * 1024 * 1024 // 1 GB
const val FILE_HEADER_LEN = 49
val MAGIC0 = 0xD1.toByte()
val MAGIC1 = 0x0C.toByte()
val FILE_MAGIC = byteArrayOf(0x44, 0x43, 0x46, 0x32) // "DCF2"

data class FrameHeader(
    val sessionId: Int,
    val seq: Int,
    val k: Int,
    val blockLen: Int,
    val totalLen: Int,
    val payloadFnv: Long
)

data class PackedOpticalFile(
    val container: ByteArray,
    val useGzip: Boolean,
    val originalSize: Int,
    val transmittedSize: Int
)

data class OpticalFile(
    val name: String,
    val type: String,
    val bytes: ByteArray,
    val sha256: ByteArray,
    val useGzip: Boolean,
    val transmittedSize: Int
)

object Protocol {

    /** FNV-1a 32-bit hash matching protocol.ts */
    fun fnv1a(bytes: ByteArray): Long {
        var h = 0x811c9dc5L
        for (b in bytes) {
            val v = (b.toInt() and 0xFF).toLong()
            h = h xor v
            h = (h * 0x01000193L) and 0xFFFFFFFFL
        }
        return h
    }

    /** Compute SHA-256 digest */
    fun sha256(bytes: ByteArray): ByteArray {
        val md = MessageDigest.getInstance("SHA-256")
        return md.digest(bytes)
    }

    fun safeFileName(name: String): String {
        val base = name.split(Regex("[/\\\\]")).lastOrNull() ?: ""
        val cleaned = base.replace(Regex("[\\u0000-\\u001f\\u007f]"), "").trim()
        return if (cleaned.isEmpty() || cleaned == "." || cleaned == "..") "transfer.bin" else cleaned
    }

    /** Packs FrameHeader + block into 20-byte Little Endian header frame */
    fun packFrame(h: FrameHeader, block: ByteArray): ByteArray {
        val out = ByteArray(HEADER_LEN + block.size)
        out[0] = MAGIC0
        out[1] = MAGIC1
        val buffer = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)
        buffer.position(2)
        buffer.putShort(h.sessionId.toShort())
        buffer.putInt(h.seq)
        buffer.putShort(h.k.toShort())
        buffer.putShort(h.blockLen.toShort())
        buffer.putInt(h.totalLen)
        buffer.putInt(h.payloadFnv.toInt())
        System.arraycopy(block, 0, out, HEADER_LEN, block.size)
        return out
    }

    /** Parses binary frame bytes into FrameHeader & block */
    fun parseFrame(bytes: ByteArray): Pair<FrameHeader, ByteArray>? {
        if (bytes.size <= HEADER_LEN) return null
        if (bytes[0] != MAGIC0 || bytes[1] != MAGIC1) return null

        val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
        buffer.position(2)
        val sessionId = buffer.short.toInt() and 0xFFFF
        val seq = buffer.int
        val k = buffer.short.toInt() and 0xFFFF
        val blockLen = buffer.short.toInt() and 0xFFFF
        val totalLen = buffer.int
        val payloadFnv = buffer.int.toLong() and 0xFFFFFFFFL

        if (k == 0 || blockLen == 0 || totalLen == 0) return null
        if (bytes.size != HEADER_LEN + blockLen) return null

        val header = FrameHeader(sessionId, seq, k, blockLen, totalLen, payloadFnv)
        val block = ByteArray(blockLen)
        System.arraycopy(bytes, HEADER_LEN, block, 0, blockLen)

        return Pair(header, block)
    }

    /** Packs raw file into DCF2 container format matching protocol.ts */
    fun packFile(name: String, type: String, bytes: ByteArray): PackedOpticalFile {
        require(bytes.isNotEmpty()) { "Choose a non-empty file." }
        val cleanName = safeFileName(name)
        val cleanType = if (type.isBlank()) "application/octet-stream" else type

        val nameBytes = cleanName.toByteArray(Charsets.UTF_8)
        val typeBytes = cleanType.toByteArray(Charsets.UTF_8)

        var tryGzip = bytes.size >= 768
        var transmitted = bytes
        var useGzip = false

        if (tryGzip) {
            try {
                val bos = ByteArrayOutputStream()
                GZIPOutputStream(bos).use { gzos -> gzos.write(bytes) }
                val compressed = bos.toByteArray()
                if (compressed.size + 64 < bytes.size) {
                    transmitted = compressed
                    useGzip = true
                }
            } catch (_: Exception) {}
        }

        val digestBytes = sha256(bytes)
        val out = ByteArray(FILE_HEADER_LEN + nameBytes.size + typeBytes.size + transmitted.size)
        System.arraycopy(FILE_MAGIC, 0, out, 0, 4)

        val buffer = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)
        buffer.position(4)
        buffer.put(if (useGzip) 1.toByte() else 0.toByte())
        buffer.putShort(nameBytes.size.toShort())
        buffer.putShort(typeBytes.size.toShort())
        buffer.putInt(bytes.size)
        buffer.putInt(transmitted.size)
        System.arraycopy(digestBytes, 0, out, 17, 32)

        var offset = FILE_HEADER_LEN
        System.arraycopy(nameBytes, 0, out, offset, nameBytes.size)
        offset += nameBytes.size
        System.arraycopy(typeBytes, 0, out, offset, typeBytes.size)
        offset += typeBytes.size
        System.arraycopy(transmitted, 0, out, offset, transmitted.size)

        return PackedOpticalFile(
            container = out,
            useGzip = useGzip,
            originalSize = bytes.size,
            transmittedSize = transmitted.size
        )
    }

    /** Unpacks DCF2 container payload */
    fun unpackFile(container: ByteArray): OpticalFile {
        require(container.size >= FILE_HEADER_LEN) { "File header is incomplete." }
        for (i in 0 until 4) {
            require(container[i] == FILE_MAGIC[i]) { "File header is invalid." }
        }

        val buffer = ByteBuffer.wrap(container).order(ByteOrder.LITTLE_ENDIAN)
        buffer.position(4)
        val useGzip = buffer.get().toInt() == 1
        val nameLen = buffer.short.toInt() and 0xFFFF
        val typeLen = buffer.short.toInt() and 0xFFFF
        val fileLength = buffer.int
        val transmittedLength = buffer.int

        val sha256Expected = ByteArray(32)
        System.arraycopy(container, 17, sha256Expected, 0, 32)

        val dataOffset = FILE_HEADER_LEN + nameLen + typeLen
        require(dataOffset + transmittedLength == container.size) { "File length does not match container header." }

        val nameBytes = ByteArray(nameLen)
        System.arraycopy(container, FILE_HEADER_LEN, nameBytes, 0, nameLen)

        val typeBytes = ByteArray(typeLen)
        System.arraycopy(container, FILE_HEADER_LEN + nameLen, typeBytes, 0, typeLen)

        val transmitted = ByteArray(transmittedLength)
        System.arraycopy(container, dataOffset, transmitted, 0, transmittedLength)

        val decompressed = if (useGzip) {
            val bis = ByteArrayInputStream(transmitted)
            val gzis = GZIPInputStream(bis)
            gzis.readBytes()
        } else {
            transmitted
        }

        require(decompressed.size == fileLength) { "Decompressed file length mismatch." }

        val actualSha = sha256(decompressed)
        for (i in 0 until 32) {
            require(actualSha[i] == sha256Expected[i]) { "SHA-256 integrity check failed." }
        }

        return OpticalFile(
            name = safeFileName(String(nameBytes, Charsets.UTF_8)),
            type = String(typeBytes, Charsets.UTF_8),
            bytes = decompressed,
            sha256 = actualSha,
            useGzip = useGzip,
            transmittedSize = transmittedLength
        )
    }
}
